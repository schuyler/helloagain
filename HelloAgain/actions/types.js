'use strict'

// Contacts
export const TOGGLE_ACTIVE = 'TOGGLE_ACTIVE'

// Friends
export const UPDATE_FRIEND = 'UPDATE_FRIEND'
export const MARK_AS_CONTACTED = 'MARK_AS_CONTACTED'
