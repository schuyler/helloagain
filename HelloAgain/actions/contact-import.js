'use strict'

import Contacts from 'react-native-contacts'

import {
  CONTACTS_LOADED, CONTACT_LOAD_FAILED
} from './types';

export const contactsLoaded = (contacts) => {
  return {type: CONTACTS_LOADED, contacts: contacts}
}

export const contactLoadFailed = (error) => {
  return {type: CONTACT_LOAD_FAILED, error: error}
}

export const loadNativeContacts = () => {
  // Create a Promise that gets all native contacts and resolves on success.
  const loadContacts = new Promise((resolve, reject) => {
    Contacts.getAll((err, contacts) => {
      err ? reject(err) : resolve(contacts)
    })
  })
  // Return an action thunk that:
  //   1. Loads all contacts
  //   2. On success, dispatches the `contactsLoaded` action
  //   3. Then, dispatches the `writeIDsForAllFriends` to write back any new
  //      IDs to the native contact store
  //   4. On failure, dispatch the `contactLoadFailed` action
  return dispatch => {
    return loadContacts
      .then(contacts => {
        let result = dispatch(contactsLoaded(contacts))
        console.log("contactsLoaded:", Object.values(result).length)
      })
      .then(_ => { return dispatch(writeIDsForAllFriends()) })
      .catch(error => dispatch(contactLoadFailed(error)))
  }
}

const generateURL = (item) => {
  return "helloagain://to/" + item.helloAgainID
}

// Check a single contact record to see if it needs the HelloAgain URL
// written to it, and if it does, write it.
export const writeIDtoNativeContact = (item) => {
  // Never been active? Don't touch the native contact.
  if (item.isActive === undefined) {
    return
  }
  // Are any of the contact's urlAddresses a HelloAgain address? Great. Skip
  // this one.
  let matchingURLs = (item.urlAddresses || []).filter(url => url.label == "HelloAgain")
  if (matchingURLs.length > 0) {
    return
  }
  // Otherwise, generate the URL, push it onto the urlAddresses list with the
  // right label, and then update the native contact store, with a log message
  // either way.
  let url = generateURL(item)
  item.urlAddresses.push({label: "HelloAgain", urlAddress: url})
  Contacts.updateContact(item, (err, contact) => {
    if (err) {
      console.log(`Error updating ${item.givenName} ${item.familyName} ` +
                  `with URL ${url}:`, err, contact)
    } else {
      console.log(`Wrote URL to Contacts: ${url}`)
    }
  })
}

const writeIDsForAllFriends = () => {
  // Return an action thunk that is pure side-effects -- just iterate
  // over the entire set of contacts and make sure that any once-active friends
  // that don't have HelloAgain URLs associated with them get one. This
  // shouldn't really ever happen, because the `toggleActive` action takes care
  // of the records one-by-one, but I guess it can't hurt to be certain.
  return (_, getState) => {
    const state = getState()
    const friends = Object.values(state.friends).filter(f => f.isActive !== undefined)
    friends.forEach(item => writeIDtoNativeContact(item))
  }
}
