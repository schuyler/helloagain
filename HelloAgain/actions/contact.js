'use strict'

import { TOGGLE_ACTIVE } from './types';

export const toggleActive = (friend) => {
  return { type: TOGGLE_ACTIVE, friend: friend }
}
