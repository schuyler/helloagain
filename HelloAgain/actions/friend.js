'use strict'

import {
  UPDATE_FRIEND,
  MARK_AS_CONTACTED
} from './types'

export const updateFriend = (friend) => {
  return { type: UPDATE_FRIEND, friend: friend }
}

export const markAsContacted = (friend) => {
  return { type: MARK_AS_CONTACTED, friend: friend }
}
