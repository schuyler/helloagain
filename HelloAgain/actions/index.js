'use strict';

// Enumerate the dispatchable actions.
export const UPDATE_FRIEND = 'UPDATE_FRIEND'

// Helpers for dispatching actions.
export const updateFriend = (friend) => {
  return { type: UPDATE_FRIEND, friend: friend }
}
