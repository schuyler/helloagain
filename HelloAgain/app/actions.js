export const IMPORT_CONTACTS = 'IMPORT_CONTACTS';

export const UPDATE_FRIEND = 'UPDATE_FRIEND';

export const FRIEND_CONTACTED = 'FRIEND_CONTACTED';

export const UPDATE_SETTINGS = 'UPDATE_SETTINGS';

export function importContacts(contactList) {
  return { type: IMPORT_CONTACTS, contactList };
}

export function updateFriend(friend) {
  return { type: UPDATE_FRIEND, friend };
}
