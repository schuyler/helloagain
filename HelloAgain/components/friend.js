'use strict';

import React, { Component } from 'react';
import {
  Text,
  View,
  StyleSheet,
  TouchableOpacity,
  Image
} from 'react-native'
import Duration from 'durationjs'
import Portrait from './portrait'
import shared from './shared'

export default class Friend extends Component {
  constructor(props) {
    super(props)
  }

  formatDurationSince(then) {
    if (!then) {
      return "never"
    }
    let ago = true ? " ago" : ""
    let delta = Date.now() - then
    let age = new Duration(delta / 1000)
    if (age.inMonths() >= 1) {
      return Math.floor(age.inMonths()) + "m" + ago
    } else if (age.inWeeks() >= 1) {
      return Math.floor(age.inWeeks()) + "w" + ago
    } else if (age.inDays() >= 1) {
      return Math.floor(age.inDays()) + "d" + ago
    } else if (age.inHours() >= 1) {
      return Math.floor(age.inHours()) + "h" + ago
    } else if (age.inMinutes() >= 1) {
      return Math.floor(age.inMinutes()) + "m" + ago
    } else {
      return "just now"
    }
  }

  render() {
    const item = this.props.item;
    let sinceLastContact = this.formatDurationSince(item.contactedAt)
    return (
      <TouchableOpacity onPress={() => this.props.onPress(item)}>
        <View style={styles.contactRow}>
          <Portrait uri={item.thumbnailPath} style={styles.contactPicture} />
          <Text style={styles.contactName}>{item.givenName} {item.familyName}</Text>
          <Text style={styles.contactLatency}>{sinceLastContact}</Text>
        </View>
      </TouchableOpacity>
    )
  }
}

// This really wants to be refactored somehow with the contact styles
const styles = StyleSheet.create({
  contactRow: shared.row,
  contactPicture: shared.rowPicture,
  contactName: shared.rowName,
  contactLatency: {
    ...shared.rowProps,
    fontSize: shared.rowProps.fontSize - 10,
    color: "grey"
  }  
})
