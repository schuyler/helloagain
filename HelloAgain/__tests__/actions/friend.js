'use strict';

import * as actions from '../../actions/friend'

let bob = {name: "Bob"};

describe('updateFriend', () => {
  it('should return an action', () => {
    expect(
      actions.updateFriend(bob)
    ).toMatchSnapshot()
  })
})

describe('markAsContacted', () => {
  it('should return an action', () => {
    expect(
      actions.markAsContacted(bob)
    ).toMatchSnapshot()
  })
})
