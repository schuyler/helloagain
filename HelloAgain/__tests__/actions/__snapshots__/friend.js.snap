exports[`markAsContacted should return an action 1`] = `
Object {
  "friend": Object {
    "name": "Bob",
  },
  "type": "MARK_AS_CONTACTED",
}
`;

exports[`updateFriend should return an action 1`] = `
Object {
  "friend": Object {
    "name": "Bob",
  },
  "type": "UPDATE_FRIEND",
}
`;
